### PR TITLE
Karl and Claire V1.

### PR DESCRIPTION
**Pull request by dkorolev**
Hi Max!

Claire is almost done, missing dependencies and sending regular keepalive. Claire's status page is as minimalistic and useful as I hope to keep it.

Karl is not doing anything beyond collecting reports and performing a callback as requested.

Claire keepalives, Karl data model, browsing, and visualization capabilities are left on me.

Also implemented and tested:
* Custom status pages (via a `Variant<>`, as discussed).
* Custom `Makefile`-s for with and without `current_build.h`
* Standalone binaries in `examples/Karl/` spawnable and testable.

The code looks worthy for a review to my taste. Please abuse and blame relentlessly.

Thanks and safe travels!
Dima

### CODE REVIEW

File: Karl/README.md [hunk 1]
```diff
@@ -1,4 +1,60 @@
...
+* Claire communicates with Karl on a nice-to-have basis.
+  * Generally, if Karl is unreachable, the client binary with Claire remains fully functional.
+* Karl can be run in stateless mode.
+  * If a production Karl is replaced by a virgin-state one, it would be populated with the state of the fleet within the next half a minute.
+* Karl maintains the historical state of all keepalives it has received.
+  * Basic Karl reporting page is a state of the fleet over certain period of time, "last five minutes" by default.
+    * The report will be based on all keepalives received within this time period.
+  * Drilldown pages include slicing per:
+    * IP address (server view),
+    * IP address and port ("endpoint" view, the "socket" other services may depent on), per codename ("
```
Comment (mzhurovich) [inline Karl/README.md]: s/depent/depend/ + unfinished line.

File: Karl/claire.h [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    // If `strict` is true, throw if Karl can be not be reached.
+    // If `strict` is false, just start the keepalives thread.
+    std::lock_guard<std::mutex> lock(mutex_);
+    if (!register_called_) {
+      register_called_ = true;
+      status_generator_ = status_filler;
+      // Only register once.
+      if (strict) {
+        // During this call, Karl would crawl the endpoint of this service, and, if everything is successful,
+        // register this service as the running and browsable one.
```
Comment (mzhurovich) [inline Karl/claire.h]: I was thinking about slightly different logic in Claire:
 * Do nothing in ctor (speaking of your comment above).
 * On `Register()` send `strict` value to Karl to control its registration behavior. I.e. regardless of `strict` value we always do POST on Karl's URL and wait for `200 OK`.
   * If `strict == false`, Karl just stores the information about the service and responds 200.
   * If `strict == true`, Karl tries to check the availability of the Claire, and only then sends 200 or error, if the service is not accessible from its side.

What do you think?

**Comment by mzhurovich** [comment]
@dkorolev Maybe move all the test to a separate dir?

File: Karl/claire.h (hunk 2 above)
Comment (dkorolev) [inline Karl/claire.h]: I still struggle to see what's wrong with registering `.current` in Claire's constructor.

If the user does nothing, their binary would still report its service/codename/port/time skew/uptime/build_info. All of the above are both straightforward to generate/respond with, and useful for monitoring/development/babysitting purposes.

As we move to scoped `HTTP(port)` ownership, I'd expect each class claiming (exclusive) access to certain HTTP port would claim it first thing in the constructor ("fail early, fail better"). Upon making that the new standard, I see no reason to not move Claire initialization right into there, effectively by designing it as `class HTTPPortAccessor { Register(...)  ... };` and `class Claire : HTTPPortAccessor { ... }`, so the user could use `Claire` to first register HTTP endpoints, and then `.RegisterWithKarl()` itself.

Let's converge on something in the next 36 hours :-)

File: Karl/README.md (hunk 1 above)
Comment (dkorolev) [inline Karl/README.md]: :+1: 8b1c4a3

**Comment by dkorolev** [comment]
@mzhurovich You mean like this: 8b799f0 ?

**Comment by dkorolev** [comment]
@mzhurovich It's work in progress, but could you take a look at cc3ab04 pls?

File: Karl/README.md [hunk 3]
```diff
@@ -23,38 +24,46 @@ General ideas behind Karl and Claire are:
...
 * **Codename**, ex. "ABCDEF".
   A random, unique, identifier of a running service. Regenerated on each binary run, changes with restart.
 
 * **Claire status**
-  A generic status report, containing, at the very least, generic binary info, and generic runtime info with basic user data.
-  Claire status reports are sent from Claire-powered services to Karl every 20 seconds.
-
-* **Binary info**
-  Service name, build date/time, compiler/environment info, git branch and commit hash, and whether the build was performed from a vanilla branch, etc.
+  A common denominator status report, containing the binary info, generic runtime ifo, and, if the service has completed initialization and informed Claire aboutit, the runtime info for this service.
```
Comment (mzhurovich) [inline Karl/README.md]: s/ifo/info/
s/aboutit/about it/

File: Karl/README.md [hunk 4]
```diff
@@ -23,38 +24,46 @@ General ideas behind Karl and Claire are:
...
 
 * **Claire status page**
-  A page available via a `GET`/`POST` request to `localhost:${port}/.current`.
-  Returns a [generally] service-dependent response, which, however, should be JSON-parsable as a Generic Claire status.
-  A complete page (what is sent as a keepalive) is accessible via `?all`; the default page returns uptime status, and `?build` returns build status.
+  A page available via a `GET`/`POST` request to `localhost:${port}[possible_prefix]/.current`. Returns a JSON with Claire status.
+  
+  This JSON, by convention, is always parsable as a generic Claire status (no service-specific runtime data).
+  
+  This JSON and can be parsed as a Claire status with a custom user status type, as long as the latter is linked into Karl's typelist. This is the way to tailor Karl's status reporting and supervision to specific service needs.
```
Comment (mzhurovich) [inline Karl/README.md]: s/`This JSON and can`/`This JSON can`/?

File: Karl/claire.h [hunk 5]
```diff
@@ -89,42 +89,41 @@ class GenericClaire final {
...
           } else {
             CURRENT_THROW(ClaireRegistrationException(service_, route));
           }
         } catch (const current::Exception&) {
           CURRENT_THROW(ClaireRegistrationException(service_, route));
         }
       } else {
-        // In non-`strict` more, simply mark the service as `registered`.
-        registered_ = true;
+        // In non-`require_karls_confirmation` more, upon being passed in the `status_filler`,
```
Comment (mzhurovich) [inline Karl/claire.h]: s/more/mode/?

**Comment by mzhurovich** [comment]
@dkorolev Spotted couple of misprints :)

File: Karl/schema.h [hunk 6]
```diff
@@ -59,23 +59,11 @@ CURRENT_STRUCT(status) {
...
-  CURRENT_FIELD(us_start, std::chrono::microseconds);
-  CURRENT_FIELD(us_now, std::chrono::microseconds);  // Uptime is calculated by Karl, along with time skew.
-
-  CURRENT_FIELD(build, Optional<build::Info>);
+using DefaultClaireServiceStatus = default_user_status::status;
 
+// For now, Karl parses the passed in JSON twice: once as `ClaireStatus` for generic response,
+// and once as `ClaireServiceStatus<T>`, where `T` is a `Variant` containing the client-, service-side blob.
+CURRENT_STRUCT_T_DERIVED(ClaireServiceStatus, ClaireStatus) {
   CURRENT_FIELD(runtime, T);
```
Comment (mzhurovich) [inline Karl/schema.h]: Don't you want to put `Variant<>` into the field type itself?

**Comment by mzhurovich** [comment]
@dkorolev Merging as is. Let's make further modifications in another pull request.

**Pull request merged**